client/src/main/kotlin/io/spine/examples/chatspn/desktop/chat/ChatsPageModel.kt

```diff
@@ -61,36 +76,210 @@ public class ChatsPageModel {
     }
 
     /**
      * Updates the model with new messages.
      */
-    public fun updateMessages(chat: ChatId, messages: MessageList) {
+    private fun updateMessages(chat: ChatId, messages: MessageList) {
         if (chatMessagesStateMap.containsKey(chat)) {
             chatMessagesStateMap[chat]!!.value = messages
         } else {
             chatMessagesStateMap[chat] = MutableStateFlow(messages)
         }
     }
+
+    /**
+     * Converts the `ChatPreview` list to the `ChatData` list.
+     */
+    private fun List<ChatPreview>.toChatDataList(): ChatList {
+        return this.stream().map { chatPreview ->
+            val lastMessage: MessageData?
+            if (chatPreview.lastMessage.equals(MessagePreview.getDefaultInstance())) {
+                lastMessage = null
+            } else {
+                lastMessage = chatPreview.lastMessage.toMessageData();
+            }
+            ChatData(
+                chatPreview.id,
+                chatPreview.name(),
+                lastMessage
+            )
+        }.collect(Collectors.toList())
+    }
+
+    /**
+     * Converts the `MessageView` list to the `MessageData` list.
+     */
+    private fun List<MessageView>.toMessageDataList(): MessageList {
+        val users = mutableMapOf<UserId, UserProfile>();
+        val messages = this.stream().map { message ->
+            val user: UserProfile
+            if (users.containsKey(message.user)) {
+                user = users[message.user]!!
+            } else {
+                user = client.findUser(message.user)!!
+                users[message.user] = user
+            }
+            MessageData(
+                message.id,
+                user,
+                message.content,
+                message.whenPosted
+            )
+        }.collect(Collectors.toList())
+        return messages
+    }
+
+    /**
+     * Converts the `MessagePreview` to `MessageData`.
+     */
+    private fun MessagePreview.toMessageData(): MessageData {
+        return MessageData(
+            this.id,
+            client.findUser(this.user)!!,
+            this.content,
+            this.whenPosted
+        )
+    }
+
+    /**
+     * Converts the `MessageView` to `MessageData`.
+     */
+    private fun MessageView.toMessageData(): MessageData {
+        return MessageData(
+            this.id,
+            client.findUser(this.user)!!,
+            this.content,
+            this.whenPosted
+        )
+    }
+
+    /**
+     * Extracts the name of the chat.
+     */
+    private fun ChatPreview.name(): String {
+        if (this.hasGroupChat()) {
+            return this.groupChat.name
+        }
+        val creator = this.personalChat.creator
+        val member = this.personalChat.member
+        if (client.authenticatedUser!!.id.equals(creator)) {
+            return client.findUser(member)!!.name
+        }
+        return client.findUser(creator)!!.name
+    }
+
+    public class UserSearchState {
```

Comment:
Right now this is a public undocumented mutable type, which I cannot even comment on.

Please improve its quality of life. And then I will be able to review it.